routes/bridges/gassistant/index.js

```diff
@@ -0,0 +1,235 @@
+// Refer to https://developers.google.com/assistant/conversational/responses for full list of response types
+
+const userUtils = require('../../../util/user');
+const db = require('../../../util/db');
+const model = require('../../../model/user');
+const secret = require('../../../util/secret_key');
+
+var router = express.Router();
+
+class GoogleAssistantDelegate {
+    constructor(locale) {
+        this._buffer = [];
+        this._locale = locale;
+        this._requestSignin = false;
+    }
+
+    send(text, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] === 'string')
+            // If there is already a text reply immediately before, we merge text replies
+            // because Google Assistant limits at most 2 chat bubbles per turn
+            this._buffer[this._buffer.length - 1] += '\n' + text;
+        else
+            this._buffer.push(text);
+    }
+
+    sendPicture(url, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] !== 'string')
+            // If there is no text reply immediately before, we add the URL
+            // because Google Assistant requires a chat bubble to accompany an Image
+            this._buffer.push(url);
+        this._buffer.push(new Image({
+            url: url,
+            alt: url,
+        }));
+    }
+
+    sendRDL(rdl, icon) {
+        this._buffer.push(new BasicCard({
+            title: rdl.displayTitle,
+            text: rdl.displayText,
+            buttons: new Button({
+                title: rdl.displayTitle,
+                url: rdl.webCallback,
+            }),
+            image: new Image({
+                url: rdl.pictureUrl,
+                alt: rdl.pictureUrl
+            }),
+            display: 'CROPPED',
+        }));
+    }
+
+    sendChoice(idx, what, title, text) {
+        let suggestions = [];
+        // Filter out buttons more than 25 characters long
+        // since Google Assistant has a cap of 25 characters
+        if (title.length <= 25)
+            suggestions.push(title.substring(0, 25));
+        else
+            console.log(`${title} exceeds max length of 25 characters`);
+        if (suggestions.length)
+            this._buffer.push(new Suggestions(suggestions));
+    }
+
+    sendButton(title, json) {
+        let suggestions = [];
+        // Filter out buttons more than 25 characters long
+        // since Google Assistant has a cap of 25 characters
+        if (title.length <= 25)
+            suggestions.push(title.substring(0, 25));
+        else
+            console.log(`${title} exceeds max length of 25 characters`);
+        if (suggestions.length)
+            this._buffer.push(new Suggestions(suggestions));
+    }
+
+    sendLink(title, url) {
+        if (url === '/user/register') {
+            this._requestSignin = true;
+        } else {
+            this._buffer.push(new Button({
+                title: title,
+                url: url,
+            }));
```

Comment:
You'll probably want to make the URL absolute here. Also, "link" buttons are not general purpose links, they are in-app navigation (things like "configure a new device", "start oauth", etc.).
(see the docs at https://almond.stanford.edu/doc/almond-dialog-api-reference.md#link-message)

Reply:
Is there a good way for me to get the URL? Since it might vary e.g. https://almond.stanford.edu vs https://almond-dev.stanford.edu

Reply:
Config.SERVER_ORIGIN